File: 신기하/합승 택시 요금.py
def solution(n, s, a, b, fares):
    # 초기 설정
    inf = int(1e9)
    answer = inf
    matrix = list([inf for _ in range(n)] for _ in range(n))

    # 자신에게 가는 것은 비용이 들지 않음
    for i in range(n):
        matrix[i][i] = 0

    # fare 설정
    for fare in fares:
        matrix[fare[0] - 1][fare[1] - 1] = fare[2]
        matrix[fare[1] - 1][fare[0] - 1] = fare[2]

    # 플로이드 와샬 알고리즘
    # 모든 점 간의 최소 거리를 구함
    for k in range(n):  # 경유지
        for i in range(n):  # 출발지
            for j in range(i, n):  # 목적지
                if i != j:
                    result = min(matrix[i][j], matrix[i][k] + matrix[k][j])
                    matrix[i][j] = matrix[j][i] = result

    # 모든 점에서 S, A, B를 가는데 걸리는 비용을 더해서 최소값이 답
    for i in range(n):
        result = matrix[s - 1][i] + matrix[a - 1][i] + matrix[b - 1][i]
        if answer > result:
            answer = result

    return answer

File: 박민상/합승택시요금.java
package programmers;

import java.util.*;

public class 합승택시요금 {
    static final int INF = 20000001;
    public static void main(String[] args) {
        solution(6, 4, 6, 2, new int[][]
                {{4, 1, 10}, {3, 5, 24}, {5, 6, 2}, {3, 1, 41}, {5, 1, 24},{4, 6, 50},{2, 4, 66},{2, 3, 22},{1, 6, 25}});
    }

    public static int solution(int n, int s, int a, int b, int[][] fares) {
        int answer = INF;

        int[][] adj = new int[n][n];
        for(int i=0; i<n; i++){
            Arrays.fill(adj[i], INF);
        }

        for(int [] fare : fares){
            int from = fare[0]-1;
            int to = fare[1]-1;
            int f = fare[2];

            adj[from][to]=f;
            adj[to][from]=f;
        }

        for(int k=0; k<n; k++){
            for(int i=0; i<n; i++){
                for(int j=0; j<n; j++){
                    if(i==j) adj[i][j]=0;
                    adj[i][j] = Math.min(adj[i][j], adj[i][k]+adj[k][j]);
                }
            }
        }

        for(int k=0; k<n; k++){
            answer = Math.min(answer, adj[s-1][k]+adj[k][a-1]+adj[k][b-1]);
        }

        System.out.println(answer);
        return answer;
    }

}
